unit/api/base_resource.py

```diff
@@ -1,34 +1,98 @@
 import json
-from typing import Optional, Dict
 import requests
+import backoff
+from typing import Optional, Dict
 from unit.models.codecs import UnitEncoder
 
+retries = 0
+
+
+def backoff_idempotency_key_handler(e):
+    return backoff_handler(e) and idempotency_key_is_present(e)
+
+
+def backoff_handler(e):
+    code = e.status_code
+    return is_timeout(code) or is_rete_limit(code) or is_server_error(code)
+
+
+def is_timeout(code):
+    return code == 408
+
+
+def is_rete_limit(code):
+    return code == 429
+
+
+def is_server_error(code):
+    return 500 <= code <= 599
+
+
+def idempotency_key_is_present(e):
+    body = json.loads(e.request.body)
+    if body is None:
+        return False
+
+    return body["data"]["attributes"].get("idempotencyKey") is not None
+
 
 class BaseResource(object):
-    def __init__(self, api_url, token):
+    def __init__(self, api_url, token, retries_amount):
+        global retries
```

Comment:
not sure about that.. if we are rewriting it always with reties amount what's the point? I think I'm missing something..

Reply:
It's not possible to use self.retries in @backoff.on_predicate